### DISCUSSION TITLE
mock prisma with vitest on next.js

### DISCUSSION THREAD

**Discussion by equiman** [prisma/prisma#20244] [Q&A] [answered]
I've followed this blog post https://www.prisma.io/blog/testing-series-1-8eRB5p0Y8o to config unit test for Prisma on a Next.js project with vitest.

# Versions

```
{
  "dependencies": {
    "@prisma/client": "5.0.0",
    "next": "13.4.10",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "devDependencies": {
    "jsdom": "22.1.0",
    "prisma": "5.0.0",
    "vitest": "0.33.0",
    "vitest-mock-extended": "1.1.4"
  }
}
```

# Postman

Running the project and debugging with Next.js I can watch Prisma

![image](https://github.com/prisma/prisma/assets/933393/ab5ebb48-9785-4b4a-8b99-0aa27d4159b1)

And get the result from Mongo 
![image](https://github.com/prisma/prisma/assets/933393/573d3d75-7954-459a-b4a8-3f1d28c779db)

# Mock Intellisense

`prismaMock` seems to be implemented correctly because IntelliSense is working as expected.

![image](https://github.com/prisma/prisma/assets/933393/c2306d15-2daa-49c0-b425-f64720297bfa)
![image](https://github.com/prisma/prisma/assets/933393/a077dbbb-7361-4732-9f08-5c3d71f05c2b)
![image](https://github.com/prisma/prisma/assets/933393/2d10b5fb-8fc3-47bb-8769-59cac7f40eb8)

# Vitest

But on test mode, I can't watch the mocked prisma variable when I run the test `npm run test`.

![image](https://github.com/prisma/prisma/assets/933393/ddce0813-570c-407c-b564-6fb4d2d408c0)

And throws an error: `TypeError: Cannot read properties of undefined (reading 'users')`

![image](https://github.com/prisma/prisma/assets/933393/53afb0cb-302c-4db2-91a3-a187ce5e3746)

It's like the same error before running the command `npx prisma generate`, but I can't figure out if I'm missing something in the configuration at vitest related to the mocking prisma client.

I've also tried renaming `prismaMock` as `prisma` but I get exactly the same result.

# Sample Project

I've created a sample project with a minimal configuration to reproduce the error. It will be greater if someone can help me to take a look at it, maybe I'm missing something.

https://github.com/deinsandbox/next-prisma-unit-test

# Other Resources I've tried

- https://github.com/eratio08/vitest-mock-extended/issues/50
- https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices
- https://github.com/prisma/prisma/issues/20146
- https://github.com/prisma/prisma/discussions/19048

**Comment by equiman** [comment]
I've added a sample project with minimal configurations to reproduce the error.

https://github.com/deinsandbox/next-prisma-unit-test

Clone and run this commands:

```bash
npm ci
npm run prisma:client
npm run test
```

**Comment by janpio** [comment]
Do I understand correctly that when running the tests via Vitest because the mocked `PrismaClient` instance is missing the models?

**Reply by equiman** [reply]
Yes, it seems to be that the mocked instance has been lost when running vitest.

**Comment by equiman** [comment]
I've tried updating  the packages versions, but still showing the same error:

```json
{
  "dependencies": {
    "@prisma/client": "5.2.0",
    "next": "13.4.19",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "devDependencies": {
    "jsdom": "22.1.0",
    "prisma": "5.2.0",
    "vitest": "0.34.3",
    "vitest-mock-extended": "1.1.4"
  }
}
```

![image](https://github.com/prisma/prisma/assets/933393/98297414-f6b7-4844-9046-b94403c74ee5)

**Comment by DitisAlex** [comment]
I'm running into the same issue, have you ever managed to fix it?

**Comment by equiman** [comment]
No, I'm trying with newer versions, but still happens the same.

**Comment by WarlockJa** [comment, marked as answer]
Hello. I've encountered the same issue and I managed to solve it by modifying two things. First I've made mock prisma instance a non default export (@/lib/__mocks__/globalForPrisma): 
```
import { PrismaClient } from "@prisma/client";
import { beforeEach } from "vitest";
import { mockReset, mockDeep } from "vitest-mock-extended";

beforeEach(() => {
  mockReset(prisma);
});

const prisma = mockDeep<PrismaClient>();
export { prisma };
```
and second, in my test, I've modified vi.mock to look like this
```
vi.mock("@/lib/globalForPrisma", async () => {
  const actual = await vi.importActual("@/lib/__mocks__/globalForPrisma");
  return {
    ...actual,
  };
});
```
according to the [vitest docs](https://vitest.dev/api/vi.html#vi-mock). With this mock prisma instance was no longer undefined.

**Reply by equiman** [reply]
Finally works, many thanks!

If some one are using TS that's the way.

```ts
import { it, expect, describe, vi } from 'vitest'
import { prisma as prismaMock } from '@/libs/__mocks__/prisma'
// vi.mock('@/libs/__mocks__/prisma')

vi.mock("@/libs/__mocks__/prisma", async () => {
  const actual = await vi.importActual<typeof import('@/libs/__mocks__/prisma')>("@/libs/__mocks__/prisma")
  return {
    ...actual,
  };
});

// Tests.....
```
